<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dmplot">
<title>Review differential expression analysis (DEA) and volcano plots • dmplot</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Review differential expression analysis (DEA) and volcano plots">
<meta property="og:description" content="dmplot">
<meta property="og:image" content="https://dereckmezquita.github.io/dmplot/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dmplot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.8.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/fin_getting-started-with-the-dmplot-framework-for-finance.html">Getting Started with the dmplot framework for finance</a>
    <a class="dropdown-item" href="../articles/fin_intro-to-financial-indicators-with-dmplot.html">Intro to Financial Indicators with dmplot</a>
    <a class="dropdown-item" href="../articles/fin_cpp-high-performance-technical-indicators.html">C++ high-performance technical indicators</a>
    <a class="dropdown-item" href="../articles/fin_intro-to-monte-carlo-simulations.html">Introduction to Monte Carlo Simulations</a>
    <a class="dropdown-item" href="../articles/bio_review-differential-expression-analysis-volcano-plots.html">Review differential expression analysis (DEA) and volcano plots</a>
    <a class="dropdown-item" href="../articles/stat_the-comparison-class-for-statistics.html">The Comparison class for statistics</a>
    <a class="dropdown-item" href="../articles/stat_review-principal-component-analysis.html">Review of Principal Component Analysis (PCA)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/donate.html">Donate</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dereckmezquita/dmplot" target="_blank">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.youtube.com/@derecksyoutube" target="_blank">
    <span class="fa fa-youtube fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://derecksnotes.com" target="_blank">derecksnotes.com</a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Review differential expression analysis (DEA) and volcano plots</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dereckmezquita/dmplot/blob/HEAD/vignettes/bio_review-differential-expression-analysis-volcano-plots.Rmd" class="external-link"><code>vignettes/bio_review-differential-expression-analysis-volcano-plots.Rmd</code></a></small>
      <div class="d-none name"><code>bio_review-differential-expression-analysis-volcano-plots.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction-and-context">Introduction and context<a class="anchor" aria-label="anchor" href="#introduction-and-context"></a>
</h2>
<p>Differential expression analysis (DEA) is a fundamental technique in
bioinformatics, particularly in the field of transcriptomics. This
review will cover the key concepts, statistical methods, and
visualisation techniques used in differential expression analysis, with
a focus on RNA sequencing (RNA-seq) data.</p>
</div>
<div class="section level2">
<h2 id="volcano-plots-with-dmplot">Volcano plots with <code>dmplot</code><a class="anchor" aria-label="anchor" href="#volcano-plots-with-dmplot"></a>
</h2>
<p>Simply put a volcano plot is a scatter plot that combines statistical
significance (y-axis) with the magnitude of change (x-axis) in a feature
(gene, transcript, etc.) between two conditions. The plot is named after
its characteristic shape, with significant features forming peaks that
resemble a volcano.</p>
<p>This is typically used in differential expression analysis in
bioinformatics but has other applications as well.</p>
<p><code>dmplot</code> makes creating a volcano plot easy with its
<code>Volcano</code> R6 class. You can do it in 3 simple steps:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">box</span><span class="fu">::</span><span class="fu"><a href="http://klmr.me/box/reference/use.html" class="external-link">use</a></span><span class="op">(</span>dt <span class="op">=</span> <span class="va">data.table</span><span class="op">)</span></span>
<span><span class="fu">box</span><span class="fu">::</span><span class="fu"><a href="http://klmr.me/box/reference/use.html" class="external-link">use</a></span><span class="op">(</span><span class="va">dmplot</span><span class="op">[</span> <span class="va">Volcano</span> <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 1. load data</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">diff_expr_res</span>, package <span class="op">=</span> <span class="st">"dmplot"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diff_expr_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;       feature     log2FC  p_value      fdr</span></span>
<span><span class="co">#&gt;        &lt;char&gt;      &lt;num&gt;    &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: nCvahjxZZe -4.4653827 1.84e-12 2.95e-08</span></span>
<span><span class="co">#&gt; 2: xokTmQulss -4.1254298 2.77e-10 2.23e-06</span></span>
<span><span class="co">#&gt; 3: EsqEEnrrMA -0.7639582 6.92e-10 3.70e-06</span></span>
<span><span class="co">#&gt; 4: MesqnUNFSM -1.3692713 1.79e-09 7.20e-06</span></span>
<span><span class="co">#&gt; 5: vHRmtdhRnW -0.6481394 3.48e-09 1.12e-05</span></span>
<span><span class="co">#&gt; 6: HHvlskYCEL -3.1067641 9.15e-09 2.45e-05</span></span>
<span></span>
<span><span class="co"># 2. create the Volcano object</span></span>
<span><span class="va">volc</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Volcano.html">Volcano</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>data <span class="op">=</span> <span class="va">diff_expr_res</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># 3. create the plot</span></span>
<span><span class="va">volcano_plot</span> <span class="op">&lt;-</span> <span class="va">volc</span><span class="op">$</span><span class="fu">plot_volcano</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">volcano_plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="bio_review-differential-expression-analysis-volcano-plots_files/figure-html/load-data-1.png" width="1800" style="display: block; margin: auto;"></p>
<p>Continue reading for a review of differential expression analysis and
volcano plots.</p>
</div>
<div class="section level2">
<h2 id="overview-of-differential-expression-analysis">Overview of Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#overview-of-differential-expression-analysis"></a>
</h2>
<p>Differential expression analysis aims to identify genes or
transcripts that show significant differences in expression levels
between two or more biological conditions. This technique is crucial for
understanding:</p>
<ul>
<li>Gene regulation mechanisms</li>
<li>Cellular responses to stimuli or treatments</li>
<li>Differences between cell types or disease states</li>
</ul>
<p>In order to execute DEA we must:</p>
<ol style="list-style-type: decimal">
<li>Design an experimenta</li>
<li>Collect and process RNA-seq data</li>
<li>Perform statistical analysis</li>
</ol>
<div class="section level3">
<h3 id="rna-seq-data">RNA-seq Data<a class="anchor" aria-label="anchor" href="#rna-seq-data"></a>
</h3>
<p>Before statistical analysis, raw RNA-seq data undergoes several
processing steps:</p>
<ol style="list-style-type: decimal">
<li>Quality control</li>
<li>Read alignment or pseudoalignment</li>
<li>Quantification of gene/transcript expression</li>
</ol>
<p>The output is typically a count matrix, where rows represent
genes/transcripts and columns represent samples.</p>
</div>
<div class="section level3">
<h3 id="statistics">Statistics<a class="anchor" aria-label="anchor" href="#statistics"></a>
</h3>
<div class="section level4">
<h4 id="normalisation">Normalisation<a class="anchor" aria-label="anchor" href="#normalisation"></a>
</h4>
<p>Raw count data needs to be normalised to account for differences in
sequencing depth and other technical biases. Common methods include:</p>
<ul>
<li>TPM (Transcripts Per Million) - not for differential expression
analysis<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;FPKM is not recommended due to its sensitivity to
outliers and lack of proper normalisation for library size. &lt;a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html" class="external-link"&gt;Teaching
materials at the Harvard Chan Bioinformatics Core&lt;/a&gt; &lt;a href="https://github.com/hbctraining/DGE_workshop" class="external-link"&gt;DGE_workshop&lt;/a&gt;&lt;/p&gt;'><sup>1</sup></a>
</li>
<li>RPKM/FPKM (Reads/Fragments Per Kilobase Million) - no longer
recommended<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;FPKM is not recommended due to its sensitivity to
outliers and lack of proper normalisation for library size. &lt;a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html" class="external-link"&gt;Teaching
materials at the Harvard Chan Bioinformatics Core&lt;/a&gt; &lt;a href="https://github.com/hbctraining/DGE_workshop" class="external-link"&gt;DGE_workshop&lt;/a&gt;&lt;/p&gt;'><sup>2</sup></a>
</li>
<li>DESeq2’s median of ratios method - “gene count comparisons between
samples and for DE analysis; NOT for within sample comparisons”<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;FPKM is not recommended due to its sensitivity to
outliers and lack of proper normalisation for library size. &lt;a href="https://hbctraining.github.io/DGE_workshop/lessons/02_DGE_count_normalization.html" class="external-link"&gt;Teaching
materials at the Harvard Chan Bioinformatics Core&lt;/a&gt; &lt;a href="https://github.com/hbctraining/DGE_workshop" class="external-link"&gt;DGE_workshop&lt;/a&gt;&lt;/p&gt;'><sup>3</sup></a>
</li>
<li>EdgeR’s TMM (Trimmed Mean of M-values) method</li>
</ul>
</div>
<div class="section level4">
<h4 id="modelling-gene-expression">Modelling Gene Expression<a class="anchor" aria-label="anchor" href="#modelling-gene-expression"></a>
</h4>
<p>Gene expression is often modelled using a negative binomial
distribution:</p>
<p><span class="math display">\[ Y_{ij} \sim NB(\mu_{ij}, \alpha_i)
\]</span></p>
<p>Where:</p>
<ul>
<li>
<span class="math inline">\(Y_{ij}\)</span> is the count for gene i
in sample j</li>
<li>
<span class="math inline">\(\mu_{ij}\)</span> is the mean</li>
<li>
<span class="math inline">\(\alpha_i\)</span> is the dispersion
parameter</li>
</ul>
<p>The mean <span class="math inline">\(\mu_{ij}\)</span> is related to
the experimental conditions through a log-linear model:</p>
<p><span class="math display">\[ \log(\mu_{ij}) = x_j'\beta_i
\]</span></p>
<p>Where <span class="math inline">\(x_j\)</span> is a vector of
covariates and <span class="math inline">\(\beta_i\)</span> is a vector
of regression coefficients.</p>
</div>
<div class="section level4">
<h4 id="hypothesis-testing---the-statistical-analysis">Hypothesis Testing - the Statistical Analysis<a class="anchor" aria-label="anchor" href="#hypothesis-testing---the-statistical-analysis"></a>
</h4>
<p>For each gene, we test the null hypothesis that there is no
difference in expression between conditions. This typically
involves:</p>
<ol style="list-style-type: decimal">
<li>Estimating model parameters</li>
<li>Calculating a test statistic (e.g., Wald test, likelihood ratio
test)</li>
<li>Computing a p-value</li>
</ol>
<p>Simply put what this means is that we go over every row of the matrix
and execute a statistical test to determine if the gene is
differentially expressed between the two conditions.</p>
<p>We end up with a result table that contains the gene name, the
p-value, and the log fold change. It will look a little like the
following:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diff_expr_res</span><span class="op">[</span>, <span class="op">-</span><span class="st">"fdr"</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt;       feature     log2FC  p_value</span></span>
<span><span class="co">#&gt;        &lt;char&gt;      &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: nCvahjxZZe -4.4653827 1.84e-12</span></span>
<span><span class="co">#&gt; 2: xokTmQulss -4.1254298 2.77e-10</span></span>
<span><span class="co">#&gt; 3: EsqEEnrrMA -0.7639582 6.92e-10</span></span>
<span><span class="co">#&gt; 4: MesqnUNFSM -1.3692713 1.79e-09</span></span>
<span><span class="co">#&gt; 5: vHRmtdhRnW -0.6481394 3.48e-09</span></span>
<span><span class="co">#&gt; 6: HHvlskYCEL -3.1067641 9.15e-09</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="false-discovery-rate-fdr">False Discovery Rate (FDR)<a class="anchor" aria-label="anchor" href="#false-discovery-rate-fdr"></a>
</h4>
<p>FDR (false discovery rate) is a tool used to eliminate bad results
which could look good. This is used for determining the statistical
significance of the potentially differentially expressed genes. This is
done by adjusting the p-values with the FDR method.</p>
<p>In the example of a volcano plot p-values are sometimes used on this
y-axis for determining if a gene is statistically significant or not. A
p-value cut-off of 0.05 is often used, meaning that 95% of the time you
will get results from samples that overlap in their distributions -
sometimes we get samples that do not overlap. A gene could have a
p-value less than 0.05 yet have it be a false positive.</p>
<p>FDR is a tool used in the Benjamini-Hochberg method. This method is
more stringent and helps to eliminate these false positive
discoveries.</p>
<p>When assessing p-values what we want to look for is something called
an anti-conservative distribution in p-values. P-values for genes that
are not statistically significant should be spread evenly between 0 and
1, when there is an experimental condition that may skew expression
levels there tends to be a large number p-values accumulating at under
0.05 (5%).</p>
<p>Typically one would look at these distributions by eye to determine
an anti-conservative distribution and a cut-off. Benjamini-Hochberg have
formalised this procedure into a formula; this allows us to adjust
p-values to eliminate false positives that may have been otherwise
reported as significant.</p>
<p>Thus after adjusting the p-values of our data we have a new column on
our data:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">diff_expr_res</span><span class="op">)</span></span>
<span><span class="co">#&gt;       feature     log2FC  p_value      fdr</span></span>
<span><span class="co">#&gt;        &lt;char&gt;      &lt;num&gt;    &lt;num&gt;    &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1: nCvahjxZZe -4.4653827 1.84e-12 2.95e-08</span></span>
<span><span class="co">#&gt; 2: xokTmQulss -4.1254298 2.77e-10 2.23e-06</span></span>
<span><span class="co">#&gt; 3: EsqEEnrrMA -0.7639582 6.92e-10 3.70e-06</span></span>
<span><span class="co">#&gt; 4: MesqnUNFSM -1.3692713 1.79e-09 7.20e-06</span></span>
<span><span class="co">#&gt; 5: vHRmtdhRnW -0.6481394 3.48e-09 1.12e-05</span></span>
<span><span class="co">#&gt; 6: HHvlskYCEL -3.1067641 9.15e-09 2.45e-05</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="log-fold-change">Log Fold Change<a class="anchor" aria-label="anchor" href="#log-fold-change"></a>
</h3>
<p><code>log2FC</code> stands for log fold change - mathematically this
represents: <code>log2(fold change)</code>. A fold change is calculated
as a change in expression relative to a control. If a gene is expressed
5X (5X FC) as much in the test sample as the control the FC is 5, and
its <code>log2FC</code> is about 2.32.</p>
<p>Typically a an absolute change of 1.5 <code>log2FC</code> relative to
the control as a cut-off is used to determine if a gene is
differentially expressed or not.</p>
<p>So a log2FC of 1.5 would actually represent a
<code>2^{1.5} = 2.82 FC</code> increase relative to the control. A
cut-off of 1.5 is commonly used in these kind of analysis - and can be
adjusted as per results/research in question.</p>
<p>Why use <code>log2FC</code> vs FC? The raw FC values when calculated
vary from 0 to positive infinity. They are centred around 1. Meaning
that a FC of 1 means the gene in the test individual was equal in level
of expression to the control. Downward regulation or expression in a
negative direction relative to the control are squeezed between 0 and 1,
while the positive/upregulation fold changes can vary from 1 to
infinity.</p>
<p>Using a <code>log2FC</code> allows us to recentre the values around 0
and express them symmetrically for both positive and negative FCs.</p>
<p>The log2 fold change for a gene i between two conditions is
calculated as:</p>
<p><span class="math display">\[
\log_2FC_i = \log_2\left(\frac{\text{Expression in condition
2}_i}{\text{Expression in condition 1}_i}\right)
\]</span></p>
<p>Notice on a volcano plot the x-axis is centred around 0. This is
because the log fold change is symmetrical around 0. A log fold change
of 0 means no difference in expression between the two conditions.</p>
</div>
</div>
<div class="section level2">
<h2 id="visualisation-with-volcano-plots">Visualisation with Volcano Plots<a class="anchor" aria-label="anchor" href="#visualisation-with-volcano-plots"></a>
</h2>
<p>Volcano plots are a popular way to visualise the results of
differential expression analysis. They combine statistical significance
with the magnitude of change.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">box</span><span class="fu">::</span><span class="fu"><a href="http://klmr.me/box/reference/use.html" class="external-link">use</a></span><span class="op">(</span><span class="va">dmplot</span><span class="op">[</span> <span class="va">theme_dereck_light</span> <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># since dmplot always returns a ggplot object we can make adjustments</span></span>
<span><span class="va">volcano_plot2</span> <span class="op">&lt;-</span> <span class="va">volcano_plot</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="../reference/theme_dereck_light.html">theme_dereck_light</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">volcano_plot2</span><span class="op">)</span></span></code></pre></div>
<p><img src="bio_review-differential-expression-analysis-volcano-plots_files/figure-html/volcano-plot-1.png" width="1800" style="display: block; margin: auto;"></p>
<div class="section level3">
<h3 id="interpreting-the-volcano-plot">Interpreting the Volcano Plot<a class="anchor" aria-label="anchor" href="#interpreting-the-volcano-plot"></a>
</h3>
<p>In the resulting plot:</p>
<ul>
<li>X-axis: log2 fold change</li>
<li>Y-axis: -log10(FDR)</li>
<li>Red points: Significantly up-regulated genes (FDR &lt; 0.05 and
log2FC &gt; 1)</li>
<li>Blue points: Significantly down-regulated genes (FDR &lt; 0.05 and
log2FC &lt; -1)</li>
<li>Black points: Non-significant changes</li>
<li>Labelled points: Top 15 most significant genes</li>
</ul>
<p>The plot allows quick identification of genes with both large fold
changes and high statistical significance.</p>
</div>
</div>
<div class="section level2">
<h2 id="biological-interpretation">Biological Interpretation<a class="anchor" aria-label="anchor" href="#biological-interpretation"></a>
</h2>
<p>While statistical analysis identifies differentially expressed genes,
biological interpretation is crucial. This may involve:</p>
<ol style="list-style-type: decimal">
<li>Gene Ontology (GO) enrichment analysis</li>
<li>Pathway analysis</li>
<li>Network analysis</li>
<li>Integration with other omics data</li>
</ol>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>Differential expression analysis is a powerful tool in
bioinformatics, allowing researchers to identify genes that respond to
experimental conditions or differ between biological states. By
combining rigorous statistical methods with intuitive visualisation
techniques like volcano plots, it provides valuable insights into gene
regulation and cellular function.</p>
<p><code>dmplot</code>’s <code>Volcano</code> R6 class presented here
offers a standardised and quick way way to create volcano plots,
facilitating the exploration and presentation of differential expression
results. However, it’s important to remember that these plots are just
one step in the broader process of understanding biological systems
through transcriptomic data.</p>
</div>
<div class="section level2">
<h2 id="further-reading-and-related-analyses">Further Reading and Related Analyses<a class="anchor" aria-label="anchor" href="#further-reading-and-related-analyses"></a>
</h2>
<p>While differential expression analysis provides valuable insights
into individual gene behaviour, researchers often seek to understand
broader patterns and biological contexts. Here are some related analyses
you might find interesting:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Gene Set Enrichment Analysis (GSEA)</strong>: Identifies
gene sets that are significantly enriched in different conditions or
phenotypes.</li>
<li>
<strong>Weighted Gene Co-expression Network Analysis
(WGCNA)</strong>: Identifies modules of highly correlated genes and
relates them to external sample traits.</li>
<li>
<strong>Time-course Expression Analysis</strong>: Examines how gene
expression changes over time, useful for developmental studies or
response to treatments.</li>
<li>
<strong>Single-cell RNA-seq Analysis</strong>: Extends differential
expression analysis to the single-cell level, revealing
cell-type-specific responses and heterogeneity within populations.</li>
<li>
<strong>Functional Annotation Clustering</strong>: Groups similar
annotations together to reduce redundancy and highlight biological
themes.</li>
</ol>
<p>These advanced analyses can provide deeper biological insights and
complement the results of standard differential expression analysis.
They often require specialised tools and careful interpretation but can
greatly enhance our understanding of complex biological systems.</p>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://derecksnotes.com" class="external-link">Dereck Mezquita</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
