<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="dmplot">
<title>Review of Principal Component Analysis (PCA) • dmplot</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Review of Principal Component Analysis (PCA)">
<meta property="og:description" content="dmplot">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">dmplot</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.8.0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/fin_getting-started-with-the-dmplot-framwork-for-finance.html">Getting Started with the dmplot framework for finance</a>
    <a class="dropdown-item" href="../articles/fin_intro-to-financial-indicators-with-dmplot.html">Intro to Financial Indicators with dmplot</a>
    <a class="dropdown-item" href="../articles/fin_cpp-high-performance-technical-indicators.html">C++ high-performance technical indicators</a>
    <a class="dropdown-item" href="../articles/fin_intro-to-monte-carlo-simulations.html">Introduction to Monte Carlo Simulations</a>
    <a class="dropdown-item" href="../articles/bio_review-differential-expression-analysis-volcano-plots.html">Review differential expression analysis (DEA) and volcano plots</a>
    <a class="dropdown-item" href="../articles/stat_the-comparison-class-for-statistics.html">The Comparison class for statistics</a>
    <a class="dropdown-item" href="../articles/stat_review-principal-component-analysis.html">Review of Principal Component Analysis (PCA)</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../articles/donate.html">Donate</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/dereckmezquita/dmplot" target="_blank">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://www.youtube.com/@derecksyoutube" target="_blank">
    <span class="fa fa-youtube fa-lg"></span>
     
  </a>
</li>
<li class="nav-item">
  <a class="external-link nav-link" href="https://derecksnotes.com" target="_blank">derecksnotes.com</a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Review of Principal Component Analysis (PCA)</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dereckmezquita/dmplot/blob/HEAD/vignettes/stat_review-principal-component-analysis.Rmd" class="external-link"><code>vignettes/stat_review-principal-component-analysis.Rmd</code></a></small>
      <div class="d-none name"><code>stat_review-principal-component-analysis.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Principal Component Analysis (PCA) is a powerful dimensionality
reduction technique widely used in data analysis and visualisation. This
vignette provides a comprehensive guide to understanding PCA and using
the <code>Pca</code> class implemented from <code>dmplot</code>.</p>
<p><code>dmplot</code> makes executing a relatively simple matter via
use of the <code>Pca</code> and <code>Comparison</code> (optional)
classes.</p>
<ol style="list-style-type: decimal">
<li>Prepare your data: your high-dimensional data (HDD) should be in a
data.table format with features as rows and samples as columns the first
column must be named “feature” and contain the feature names.</li>
<li>Create a <code>Comparison</code> object (optional): if you have
groups within your data that you want to compare, you can create a
<code>Comparison</code> object to represent these groups.
<ul>
<li>Using a comparison object will automatically filter your dataset to
only include the samples in the comparison and create plots that
highlight the differences between the groups.</li>
</ul>
</li>
<li>Create a <code>Pca</code> object: pass your data and comparison
object (if applicable) to the <code>Pca</code> class to perform
PCA.</li>
</ol>
<p>Now just execute <code>Pca$prcomp()</code> to perform PCA on your
data and access plots via <code>Pca$plot_scree()</code> and
<code>Pca$plot_scatter()</code>.</p>
</div>
<div class="section level2">
<h2 id="what-is-principal-component-analysis">What is Principal Component Analysis?<a class="anchor" aria-label="anchor" href="#what-is-principal-component-analysis"></a>
</h2>
<p>PCA is a statistical procedure that transforms a set of observations
of possibly correlated variables into a set of values of linearly
uncorrelated variables called principal components. This transformation
is defined in such a way that the first principal component has the
largest possible variance, and each succeeding component has the highest
variance possible under the constraint that it is orthogonal to the
preceding components.</p>
<p>What does that mean? What is orthogonal to what?</p>
<p>We are accustomed to living in a three dimensional world but in this
case we are using the word dimension simply as a way to describe a
feature on which something was measured. For example, if we have a
dataset with 1000 features (genes, financial indicators, etc.) and 100
samples (observations) then we have a 1000-dimensional dataset. PCA is a
way to reduce this 1000-dimensional dataset to a 2-dimensional dataset
(or 3-dimensional if you prefer) that captures the most important
information in the data.</p>
<p>Now what does orthogonal mean? In this context, it means that the
principal components are uncorrelated with each other. This is important
because it allows us to capture the maximum amount of variance in the
data with the fewest number of principal components.</p>
</div>
<div class="section level2">
<h2 id="the-mathematics-behind-pca">The Mathematics Behind PCA<a class="anchor" aria-label="anchor" href="#the-mathematics-behind-pca"></a>
</h2>
<p>PCA involves several mathematical steps:</p>
<ol style="list-style-type: decimal">
<li>
<p><strong>Standardisation</strong>: PCA begins with a dataset of
n-dimensions. In our implementation, genes are dimensions and samples
are observations. The data is standardised, transforming each dimension
to have a mean of 0 and a standard deviation of 1. This step is crucial
because PCA is sensitive to the relative scaling of the original
variables.</p>
<p>Mathematically, for each feature x, we compute:</p>
<p><span class="math display">\[
x_standardised = (x - μ) / σ
\]</span></p>
<p>where μ is the mean and σ is the standard deviation of the
feature.</p>
</li>
<li>
<p><strong>Covariance Matrix Computation</strong>: A covariance
matrix is computed. This matrix indicates the covariance (shared
variance) between each pair of dimensions. The covariance between
different dimensions is used to understand the correlation structure of
the original dimensions.</p>
<p>For a dataset X with m features, the covariance matrix C is computed
as:</p>
<p><span class="math display">\[
C = (1 / (n-1)) * (X^T * X)
\]</span></p>
<p>where n is the number of observations and X^T is the transpose of
X.</p>
</li>
<li>
<p><strong>Eigendecomposition</strong>: The covariance matrix is
then decomposed into its eigenvectors and eigenvalues. Each eigenvector
represents a principal component, which is a linear combination of the
original dimensions. The associated eigenvalue represents the amount of
variance explained by the principal component.</p>
<p>We solve the equation:</p>
<p><span class="math display">\[
C * v = λ * v
\]</span></p>
<p>where v is an eigenvector and λ is the corresponding eigenvalue.</p>
<p>The eigenvectors are ordered by their corresponding eigenvalues, so
the first principal component (PC1) explains the most variance, followed
by PC2, and so on.</p>
</li>
<li><p><strong>Selection of Principal Components</strong>: Depending on
the goal of the analysis, some or all of the principal components can be
selected for further analysis. The ‘elbow method’ is commonly used,
where you plot the explained variance by each principal component and
look for an ‘elbow’ in the plot as a cut-off point.</p></li>
<li><p><strong>Interpretation</strong>: The ‘top rotations’ in the
context of PCA refer to the features (genes) that contribute most to
each principal component. The ‘rotation’ matrix gives the loadings of
each feature onto each PC. By identifying features with large absolute
loadings, we can understand what features drive the separation in the
data along the principal components.</p></li>
</ol>
</div>
<div class="section level2">
<h2 id="using-the-pca-class">Using the Pca Class<a class="anchor" aria-label="anchor" href="#using-the-pca-class"></a>
</h2>
<p>The <code>Pca</code> class in <code>dmplot</code> provides methods
for each step of the PCA process, from data preparation to visualisation
of results. Here’s how to use it:</p>
<div class="section level3">
<h3 id="creating-a-pca-object">Creating a Pca Object<a class="anchor" aria-label="anchor" href="#creating-a-pca-object"></a>
</h3>
<p>First, you need to create a <code>Pca</code> object with your
data:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load required packages</span></span>
<span><span class="fu">box</span><span class="fu">::</span><span class="fu"><a href="http://klmr.me/box/reference/use.html" class="external-link">use</a></span><span class="op">(</span><span class="va">dmplot</span><span class="op">[</span><span class="va">Pca</span>, <span class="va">Comparison</span><span class="op">]</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">feature_counts</span>, package <span class="op">=</span> <span class="st">"dmplot"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># The data should be a data.table with features as rows and samples as columns</span></span>
<span><span class="co"># The first column must be named "feature" and contain the feature names</span></span>
<span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">feature_counts</span><span class="op">[</span><span class="va">GeneBiotype</span> <span class="op">==</span> <span class="st">"protein_coding"</span>, <span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"feature"</span></span>
<span></span>
<span><span class="co"># Create a Pca object</span></span>
<span><span class="va">pca_obj</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Pca.html">Pca</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>In this case we perform a generic PCA object without any comparison.
As you can see the plots do not show any groupings.</p>
<p>To execute the calculations use <code>Pca$prcomp()</code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_obj</span><span class="op">$</span><span class="fu">prcomp</span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">pca_obj</span><span class="op">)</span></span>
<span><span class="co">#&gt; PCA object:</span></span>
<span><span class="co">#&gt; -------------------------------</span></span>
<span><span class="co">#&gt; Proportion of Variance Explained by each PC:</span></span>
<span><span class="co">#&gt;         PC pct_var_explained</span></span>
<span><span class="co">#&gt;     &lt;fctr&gt;             &lt;num&gt;</span></span>
<span><span class="co">#&gt;  1:    PC1             93.51</span></span>
<span><span class="co">#&gt;  2:    PC2              3.25</span></span>
<span><span class="co">#&gt;  3:    PC3              0.78</span></span>
<span><span class="co">#&gt; ---                         </span></span>
<span><span class="co">#&gt; 70:   PC70              0.00</span></span>
<span><span class="co">#&gt; 71:   PC71              0.00</span></span>
<span><span class="co">#&gt; 72:   PC72              0.00</span></span></code></pre></div>
<p>If you want to compare groups within your data, you can provide a
<code>Comparison</code> object. Providing a <code>Comparison</code>
object will automatically filter your data to only include the samples
in the comparison and create plots that highlight samples by group.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create a comparison table</span></span>
<span><span class="va">comp_table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>   group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"B"</span>, <span class="st">"B"</span>, <span class="st">"B"</span><span class="op">)</span>,</span>
<span>   sample <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"T64552"</span>, <span class="st">"T64553"</span>, <span class="st">"T64554"</span>, <span class="st">"T64555"</span>, <span class="st">"T64546"</span>, <span class="st">"T64548"</span>, <span class="st">"T64549"</span>, <span class="st">"T64550"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a Comparison object</span></span>
<span><span class="va">comp</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Comparison.html">Comparison</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span></span>
<span>    comparison_name <span class="op">=</span> <span class="st">"A_over_B"</span>,</span>
<span>    group_order <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B"</span>, <span class="st">"A"</span><span class="op">)</span>,</span>
<span>    comparison_table <span class="op">=</span> <span class="va">comp_table</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create a Pca object with the comparison</span></span>
<span><span class="va">pca_obj2</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Pca.html">Pca</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span><span class="va">data</span>, <span class="va">comp</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_obj2</span><span class="op">$</span><span class="fu">prcomp</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>This method standardises the data, computes the covariance matrix,
performs eigendecomposition, and stores the results.</p>
</div>
<div class="section level3">
<h3 id="accessing-pca-results">Accessing PCA Results<a class="anchor" aria-label="anchor" href="#accessing-pca-results"></a>
</h3>
<p>You can access various aspects of the PCA results:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># View the raw PCA results</span></span>
<span><span class="va">pca_obj2</span><span class="op">$</span><span class="va">prcomp_results</span></span>
<span><span class="co">#&gt; Standard deviations (1, .., p=8):</span></span>
<span><span class="co">#&gt; [1] 1.138674e+06 8.773857e+04 6.518402e+04 4.366182e+04 2.366213e+04</span></span>
<span><span class="co">#&gt; [6] 2.021359e+04 1.856779e+04 5.074559e-10</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Rotation (n x k) = (21936 x 9):</span></span>
<span><span class="co">#&gt;                   feature           PC1           PC2           PC3</span></span>
<span><span class="co">#&gt;                    &lt;char&gt;         &lt;num&gt;         &lt;num&gt;         &lt;num&gt;</span></span>
<span><span class="co">#&gt;     1: ENSMUSG00000051845 -7.096769e-06 -3.435085e-06 -1.311560e-05</span></span>
<span><span class="co">#&gt;     2: ENSMUSG00000025374  2.507594e-04  2.327624e-04 -2.263580e-03</span></span>
<span><span class="co">#&gt;     3: ENSMUSG00000025609  4.829978e-04  5.629973e-04 -1.712018e-04</span></span>
<span><span class="co">#&gt;    ---                                                             </span></span>
<span><span class="co">#&gt; 21934: ENSMUSG00000063958  1.452822e-07 -1.861195e-06 -1.868383e-06</span></span>
<span><span class="co">#&gt; 21935: ENSMUSG00000096294 -3.440474e-07 -2.470131e-06 -1.668852e-06</span></span>
<span><span class="co">#&gt; 21936: ENSMUSG00000095261 -2.224891e-07  5.361425e-07  8.268138e-07</span></span>
<span><span class="co">#&gt; 5 variable(s) not shown: [PC4 &lt;num&gt;, PC5 &lt;num&gt;, PC6 &lt;num&gt;, PC7 &lt;num&gt;, PC8 &lt;num&gt;]</span></span>
<span><span class="co"># View the refined PCA results</span></span>
<span><span class="va">pca_obj2</span><span class="op">$</span><span class="va">prcomp_refined</span></span>
<span><span class="co">#&gt;        PC pct_var_explained        T64555        T64550        T64554</span></span>
<span><span class="co">#&gt;    &lt;fctr&gt;             &lt;num&gt;         &lt;num&gt;         &lt;num&gt;         &lt;num&gt;</span></span>
<span><span class="co">#&gt; 1:    PC1             98.84 -1.095036e+06  1.318588e+06 -1.009661e+06</span></span>
<span><span class="co">#&gt; 2:    PC2              0.59 -4.006736e+04 -1.002931e+05  1.444540e+04</span></span>
<span><span class="co">#&gt; 3:    PC3              0.32  2.585091e+04 -5.557073e+04  1.229583e+04</span></span>
<span><span class="co">#&gt; 4:    PC4              0.15 -6.332709e+04  1.082834e+04  2.228724e+04</span></span>
<span><span class="co">#&gt; 5:    PC5              0.04 -3.031154e+04 -1.626517e+04  3.658489e+04</span></span>
<span><span class="co">#&gt; 6:    PC6              0.03  1.610595e+04  1.165766e+04  3.180771e+04</span></span>
<span><span class="co">#&gt; 7:    PC7              0.03  1.252223e+04 -2.548565e+04 -7.162936e+03</span></span>
<span><span class="co">#&gt; 8:    PC8              0.00 -7.304681e-10  1.467666e-10 -7.570011e-10</span></span>
<span><span class="co">#&gt; 5 variable(s) not shown: [T64546 &lt;num&gt;, T64548 &lt;num&gt;, T64553 &lt;num&gt;, T64549 &lt;num&gt;, T64552 &lt;num&gt;]</span></span>
<span><span class="co"># View the top contributors to each PC</span></span>
<span><span class="va">pca_obj2</span><span class="op">$</span><span class="va">top_rotations</span></span>
<span><span class="co">#&gt; $PC1</span></span>
<span><span class="co">#&gt; ENSMUSG00000024146 ENSMUSG00000022397 ENSMUSG00000025537 </span></span>
<span><span class="co">#&gt;          0.9672892          0.1567412          0.1003282 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC2</span></span>
<span><span class="co">#&gt; ENSMUSG00000022397 ENSMUSG00000069157 ENSMUSG00000027499 </span></span>
<span><span class="co">#&gt;          0.6915537          0.3600379          0.2865109 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC3</span></span>
<span><span class="co">#&gt; ENSMUSG00000027499 ENSMUSG00000022397 ENSMUSG00000052379 </span></span>
<span><span class="co">#&gt;          0.5656072          0.4217036          0.2603293 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC4</span></span>
<span><span class="co">#&gt; ENSMUSG00000069157 ENSMUSG00000031071 ENSMUSG00000027499 </span></span>
<span><span class="co">#&gt;          0.2599749          0.2594879          0.2519562 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC5</span></span>
<span><span class="co">#&gt; ENSMUSG00000035083 ENSMUSG00000052744 ENSMUSG00000026523 </span></span>
<span><span class="co">#&gt;          0.3565922          0.2230975          0.1597955 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC6</span></span>
<span><span class="co">#&gt; ENSMUSG00000035083 ENSMUSG00000055212 ENSMUSG00000069157 </span></span>
<span><span class="co">#&gt;          0.3526967          0.2379434          0.2237858 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC7</span></span>
<span><span class="co">#&gt; ENSMUSG00000025537 ENSMUSG00000003251 ENSMUSG00000026523 </span></span>
<span><span class="co">#&gt;          0.3229935          0.2872070          0.2769927 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $PC8</span></span>
<span><span class="co">#&gt; ENSMUSG00000051845 ENSMUSG00000025916 ENSMUSG00000033093 </span></span>
<span><span class="co">#&gt;         0.99589674         0.06659177         0.05689562</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="visualising-pca-results">Visualising PCA Results<a class="anchor" aria-label="anchor" href="#visualising-pca-results"></a>
</h3>
<p>The <code>Pca</code> class provides methods for creating two common
PCA visualisations.</p>
<p>The scree plot shows the proportion of variance explained by each
principal component. This is useful for determining how many principal
components to actually contribute to any possible differences between
groups. As explained earlier the ‘elbow method’ is commonly used to
determine the number of PCs to retain. This is a point of inflection in
the scree plot where the explained variance starts to level off.</p>
<p><code>dmplot</code> provides this plot as a histogram but one can
nonetheless use the calculated values to create a line plot.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_obj2</span><span class="op">$</span><span class="fu">plot_scree</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="stat_review-principal-component-analysis_files/figure-html/pca-plots-1.png" width="1800" style="display: block; margin: auto;"></p>
<p>The scatter plot shows the samples projected onto the first two
principal components. This plot is useful for visualising the
relationships between samples in the reduced-dimensional space.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_obj2</span><span class="op">$</span><span class="fu">plot_scatter</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="stat_review-principal-component-analysis_files/figure-html/pca-scatter-plot-1.png" width="1800" style="display: block; margin: auto;"></p>
<p>Earlier we made two different PCA objects, one with a comparison and
one without. As you can see the scatter plot with the comparison shows
the samples grouped by the comparison. Below is the plot without the
comparison; notice how the samples are not coloured by groups and there
are far more samples in the plot. This is because when a
<code>Comparison</code> object is provided, the data is filtered to only
include the samples in the comparison.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_obj</span><span class="op">$</span><span class="fu">plot_scatter</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><img src="stat_review-principal-component-analysis_files/figure-html/pca-scatter-plot-no-comparison-1.png" width="1800" style="display: block; margin: auto;"></p>
</div>
</div>
<div class="section level2">
<h2 id="advanced-usage">Advanced Usage<a class="anchor" aria-label="anchor" href="#advanced-usage"></a>
</h2>
<div class="section level3">
<h3 id="customising-plots">Customising Plots<a class="anchor" aria-label="anchor" href="#customising-plots"></a>
</h3>
<p>Both <code>plot_scree()</code> and <code>plot_scatter()</code>
methods accept various parameters for customisation. For example:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scatter_plot</span> <span class="op">&lt;-</span> <span class="va">pca_obj</span><span class="op">$</span><span class="fu">plot_scatter</span><span class="op">(</span></span>
<span>   point_size <span class="op">=</span> <span class="fl">4</span>,</span>
<span>   point_alpha <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>   point_labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>show <span class="op">=</span> <span class="cn">TRUE</span>, size <span class="op">=</span> <span class="fl">3</span>, alpha <span class="op">=</span> <span class="fl">0.75</span>, font_face <span class="op">=</span> <span class="st">"bold"</span>, max_overlaps <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>,</span>
<span>   top_contributors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>show <span class="op">=</span> <span class="cn">TRUE</span>, truncate <span class="op">=</span> <span class="fl">20</span><span class="op">)</span>,</span>
<span>   title <span class="op">=</span> <span class="st">"My Custom PCA Plot"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">scatter_plot</span><span class="op">)</span></span></code></pre></div>
<p><img src="stat_review-principal-component-analysis_files/figure-html/plot-arguments-1.png" width="1800" style="display: block; margin: auto;"></p>
<p>The plotting arguments are provided for ease of use but one can
nonetheless use the <code>ggplot2</code> object to customise the plot
further.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">box</span><span class="fu">::</span><span class="fu"><a href="http://klmr.me/box/reference/use.html" class="external-link">use</a></span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span></span>
<span><span class="va">scatter_plot</span> <span class="op">+</span></span>
<span>   <span class="va">ggplot2</span><span class="op">$</span><span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>   <span class="va">ggplot2</span><span class="op">$</span><span class="fu">theme</span><span class="op">(</span></span>
<span>       plot.title <span class="op">=</span> <span class="va">ggplot2</span><span class="op">$</span><span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">16</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>       axis.title <span class="op">=</span> <span class="va">ggplot2</span><span class="op">$</span><span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">14</span><span class="op">)</span>,</span>
<span>       axis.text <span class="op">=</span> <span class="va">ggplot2</span><span class="op">$</span><span class="fu">element_text</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">12</span><span class="op">)</span></span>
<span>   <span class="op">)</span></span></code></pre></div>
<p><img src="stat_review-principal-component-analysis_files/figure-html/ggplot-customisation-1.png" width="1800" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="interpreting-results">Interpreting Results<a class="anchor" aria-label="anchor" href="#interpreting-results"></a>
</h3>
<ol style="list-style-type: decimal">
<li>
<strong>Variance Explained</strong>: Look at the
<code>pct_var_explained</code> in <code>pca_obj$prcomp_refined</code> to
understand how much of the total variance each PC explains.</li>
<li>
<strong>Sample Clustering</strong>: In the scatter plot, look for
clusters of samples. Samples that cluster together are more similar to
each other in the dimensions of the first two PCs.</li>
<li>
<strong>Top Contributors</strong>: Examine
<code>pca_obj$top_rotations</code> to understand which features
contribute most to each PC. These features are driving the differences
between your samples.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="best-practices">Best Practices<a class="anchor" aria-label="anchor" href="#best-practices"></a>
</h2>
<ol style="list-style-type: decimal">
<li>
<strong>Data Preparation</strong>: Ensure your data is properly
normalised before performing PCA. The <code>Pca</code> class expects the
data to be in a specific format (features as rows, samples as
columns).</li>
<li>
<strong>Handling Missing Data</strong>: PCA cannot handle missing
values. Ensure your data is complete or use appropriate methods to
impute missing values before PCA.</li>
<li>
<strong>Interpreting Results</strong>: Remember that PCA is an
unsupervised method. It doesn’t know about your experimental groups. If
you see separation between groups in your PCA plot, it suggests that the
largest sources of variation in your data correspond to your
experimental conditions.</li>
<li>
<strong>Number of PCs</strong>: There’s no hard rule for how many
PCs to retain. Common methods include looking for an ‘elbow’ in the
scree plot, or retaining enough PCs to explain a certain percentage
(e.g., 80%) of the total variance.</li>
<li>
<strong>Visualisation</strong>: While the first two PCs are commonly
plotted, sometimes it’s informative to look at other PCs as well,
especially if PC1 and PC2 don’t explain a large proportion of the
variance.</li>
</ol>
</div>
<div class="section level2">
<h2 id="conclusion">Conclusion<a class="anchor" aria-label="anchor" href="#conclusion"></a>
</h2>
<p>The <code>Pca</code> class provides a powerful and flexible tool for
performing Principal Component Analysis in R. By encapsulating the
entire PCA workflow - from data preparation to results visualisation -
it simplifies the process of exploring high-dimensional data.</p>
<p>Whether you’re analysing gene expression data, financial time series,
or any other high-dimensional dataset, PCA can provide valuable insights
into the structure of your data. The <code>Pca</code> class makes it
easy to leverage this powerful technique in your analyses.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://derecksnotes.com" class="external-link">Dereck Mezquita</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
